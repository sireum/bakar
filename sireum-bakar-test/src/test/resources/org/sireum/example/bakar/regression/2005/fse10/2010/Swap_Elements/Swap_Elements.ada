package Array_Utilities is
  Max_Table_Size : constant := 100;
  type Base_Index_Type is range 0..Max_Table_Size;
  subtype Index_Type is Base_Index_Type range 1..Max_Table_Size;
  type Contents_Type is range -1000 .. 1000;
  type Array_Type is array(Index_Type) of Contents_Type;

  procedure Swap(X, Y : in out Integer);
  --# derives X from Y & Y from X;
  --# post X = Y~ and Y = X~;

  procedure Swap_Elements_No_Post(T : in out Array_Type;
                                  I, J : in Index_Type);
  --# derives T from T,I,J;

  procedure Swap_Elements_weak(T : in out Array_Type;
                               I, J : in Index_Type);
  --# derives T from T,I,J;
  --# post T(I) = T~(J) and T(J) = T~(I);

  procedure Swap_Elements_strong(T : in out Array_Type;
                                 I, J : in Index_Type);
  --# derives T from T,I,J;
  --# post T = T~[I => T~(J); J => T~(I)];

end Array_Utilities;

package body Array_Utilities is

   procedure Swap(X, Y : in out Integer)
   is
      T : Integer;
   begin
      T := X; X := Y; Y := T;
   end Swap;

   procedure Swap_Elements_No_Post(T : in out Array_Type;
                               I, J : in Index_Type)
  is
    Temp : Contents_Type;
  begin
    Temp := T(I);
    T(I) := T(J);
    T(J) := Temp;
  end Swap_Elements_No_Post;

  procedure Swap_Elements_weak(T : in out Array_Type;
                               I, J : in Index_Type)
  is
    Temp : Contents_Type;
  begin
    Temp := T(I);
    T(I) := T(J);
    T(J) := Temp;
  end Swap_Elements_weak;

  procedure Swap_Elements_strong(T : in out Array_Type;
                                 I, J : in Index_Type)
  is
    Temp : Contents_Type;
  begin
    Temp := T(I);
    T(I) := T(J);
    T(J) := Temp;
  end Swap_Elements_strong;

end Array_Utilities;

