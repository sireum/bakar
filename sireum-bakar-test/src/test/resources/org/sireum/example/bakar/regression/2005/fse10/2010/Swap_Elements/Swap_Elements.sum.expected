-------------------------------------------------------------------------------
                          Semantic Analysis Summary                            
                                POGS GPL 2010                                  
            Copyright (C) 2010 Altran Praxis Limited, Bath, U.K.               
-------------------------------------------------------------------------------

Summary of:

Verification Condition files (.vcg)
Simplified Verification Condition files (.siv)
Proof Logs (.plg)
Dead Path Conjecture files (.dpc)
Summary Dead Path files (.sdp)

"status" column keys:
    1st character:
        '-' - No VC
        'S' - No SIV
        'U' - Undischarged
        'E' - Proved by Examiner
        'I' - Proved by Simplifier by Inference
        'X' - Proved by Simplifier by Contradiction
        'P' - Proved by Simplifier using User Defined Proof Rules
        'C' - Proved by Checker
        'R' - Proved by Review
        'F' - VC is False
    2nd character:
        '-' - No DPC
        'S' - No SDP
        'U' - Unchecked
        'D' - Dead path
        'L' - Live path

in the directory:
/Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Swap_Elements

Summary produced: 13-SEP-2010 14:49:12.45

File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Swap_Elements/array_utilities/swap.vcg
procedure Array_Utilities.Swap

VCs generated 13-SEP-2010 14:49:11

VCs simplified 13-SEP-2010 14:49:11

VCs for procedure_swap :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 34      | Inference          | No DPC    |   I-   |
| 2   | start | rtc check @ 34      | Inference          | No DPC    |   I-   |
| 3   | start | rtc check @ 34      | Inference          | No DPC    |   I-   |
| 4   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
 -----------------------------------------------------------------------------


File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Swap_Elements/array_utilities/swap_elements_no_post.vcg
procedure Array_Utilities.Swap_Elements_No_Post

VCs generated 13-SEP-2010 14:49:11

VCs simplified 13-SEP-2010 14:49:12

VCs for procedure_swap_elements_no_post :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 42      | Inference          | No DPC    |   I-   |
| 2   | start | rtc check @ 43      | Inference          | No DPC    |   I-   |
| 3   | start | rtc check @ 44      | Inference          | No DPC    |   I-   |
| 4   | start |    assert @ finish  | Examiner           | No DPC    |   E-   |
 -----------------------------------------------------------------------------


File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Swap_Elements/array_utilities/swap_elements_strong.vcg
procedure Array_Utilities.Swap_Elements_strong

VCs generated 13-SEP-2010 14:49:11

VCs simplified 13-SEP-2010 14:49:12

VCs for procedure_swap_elements_strong :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 62      | Inference          | No DPC    |   I-   |
| 2   | start | rtc check @ 63      | Inference          | No DPC    |   I-   |
| 3   | start | rtc check @ 64      | Inference          | No DPC    |   I-   |
| 4   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
 -----------------------------------------------------------------------------


File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Swap_Elements/array_utilities/swap_elements_weak.vcg
procedure Array_Utilities.Swap_Elements_weak

VCs generated 13-SEP-2010 14:49:11

VCs simplified 13-SEP-2010 14:49:12

VCs for procedure_swap_elements_weak :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 52      | Inference          | No DPC    |   I-   |
| 2   | start | rtc check @ 53      | Inference          | No DPC    |   I-   |
| 3   | start | rtc check @ 54      | Inference          | No DPC    |   I-   |
| 4   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
 -----------------------------------------------------------------------------


===============================================================================
Summary:

Proof strategies used by subprograms
-------------------------------------------------------------------------
Total subprograms with at least one VC proved by examiner:              1
Total subprograms with at least one VC proved by simplifier:            4
Total subprograms with at least one VC proved by contradiction:         0
Total subprograms with at least one VC proved with user proof rule:     0
Total subprograms with at least one VC proved using checker:            0
Total subprograms with at least one VC discharged by review:            0

Maximum extent of strategies used for fully proved subprograms:
-------------------------------------------------------------------------
Total subprograms with proof completed by examiner:                     0
Total subprograms with proof completed by simplifier:                   4
Total subprograms with proof completed with user defined rules:         0
Total subprograms with proof completed by checker:                      0
Total subprograms with VCs discharged by review:                        0

Overall subprogram summary:
-------------------------------------------------------------------------
Total subprograms fully proved:                                         4
Total subprograms with at least one undischarged VC:                    0
Total subprograms with at least one false VC:                           0
                                                                    -----
Total subprograms for which VCs have been generated:                    4


ZombieScope Summary:
-------------------------------------------------------------------------
Total subprograms for which DPCs have been generated:                   0
Total number subprograms with dead paths found:                         0
Total number of dead paths found:                                       0


VC summary:
-------------------------------------------------------------------------
Note: U/R denotes where the Simplifier has proved VCs using one or more user-
defined proof rules.

Total VCs by type:                                       
                            -----------Proved By Or Using------------
                     Total  Examiner Simp(U/R)  Checker Review False Undiscgd
Assert or Post:          4      1      3              0      0      0       0
Precondition check:      0      0      0              0      0      0       0
Check statement:         0      0      0              0      0      0       0
Runtime check:          12      0     12              0      0      0       0
Refinement VCs:          0      0      0              0      0      0       0
Inheritance VCs:         0      0      0              0      0      0       0
===============================================================================
Totals:                 16      1     15              0      0      0       0
% Totals:                       6%    94%             0%     0%     0%      0%
===================== End of Semantic Analysis Summary ========================
